src/main/java/com/dubture/jenkins/digitalocean/Cloud.java

```diff
@@ -230,88 +185,105 @@ private boolean addProvisionedSlave(String imageId) {
      * @return
      */
     @Override
-    public Collection<NodeProvisioner.PlannedNode> provision(Label label, int excessWorkload) {
-
-        if (label != null) {
-            LOGGER.info("Provisioning digitalocean droplet for label " + label.getName() + ", excessWorkload " + excessWorkload);
-        } else {
-            LOGGER.info("Provisioning digitalocean droplet without label, excessWorkload " + excessWorkload);
-        }
-
-        List<NodeProvisioner.PlannedNode> nodes = new ArrayList<NodeProvisioner.PlannedNode>();
-        final DigitalOceanClient apiClient = new DigitalOceanClient(authToken);
-        final SlaveTemplate template = getTemplate(label);
+    public Collection<NodeProvisioner.PlannedNode> provision(final Label label, int excessWorkload) {
+        synchronized (provisionSynchronizor) {
+            List<NodeProvisioner.PlannedNode> provisioningNodes = new ArrayList<NodeProvisioner.PlannedNode>();
+            try {
+                while (excessWorkload > 0) {
 
-        try {
-            while (excessWorkload > 0) {
-                if (!addProvisionedSlave(template.getImageId())) {
-                    break;
-                }
+                    if (isInstanceCapReached()) {
+                        LOGGER.log(Level.INFO, "Instance cap of " + getInstanceCap() + " reached, not provisioning.");
+                        break;
+                    }
 
-                final String dropletName = template.createDropletName();
-                nodes.add(new NodeProvisioner.PlannedNode(dropletName, Computer.threadPoolForRemoting.submit(new Callable<Node>() {
+                    final SlaveTemplate template = getTemplateBelowInstanceCap(label);
+                    if (template == null) {
+                        break;
+                    }
 
-                    public Node call() throws Exception {
-                        // TODO: record the output somewhere
-                        try {
-                            Slave slave = template.provision(apiClient, dropletName, privateKey, sshKeyId, new StreamTaskListener(System.out, Charset.defaultCharset()));
+                    final String dropletName = DropletName.generateDropletName(name, template.getName());
+
+                    provisioningNodes.add(new NodeProvisioner.PlannedNode(dropletName, Computer.threadPoolForRemoting.submit(new Callable<Node>() {
+                        public Node call() throws Exception {
+                            Slave slave;
+                            synchronized (provisionSynchronizor) {
```

Comment:
There is no issue because of this synchronization. A Cloud instance can indeed give Jenkins more nodes that the instance cap allows, but right before we create a droplet for a node, we check one more time if the instance cap of both the cloud and the slave template were reached, and if so, we return `null`, otherwise we ask DO to create a droplet.